core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,555 @@
+      }
+    }
+
+    delayOrExecute(new FlushEntry());
+  }
+
+  @Override
+  public boolean isReady() {
+    for (ClientStream substream : state.drainedSubstreams) {
+      if (substream.isReady()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public void setCompressor(final Compressor compressor) {
+    class CompressorEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setCompressor(compressor);
+      }
+    }
+
+    delayOrExecute(new CompressorEntry());
+  }
+
+  @Override
+  public void setFullStreamDecompression(final boolean fullStreamDecompression) {
+    class FullStreamDecompressionEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setFullStreamDecompression(fullStreamDecompression);
+      }
+    }
+
+    delayOrExecute(new FullStreamDecompressionEntry());
+  }
+
+  @Override
+  public void setMessageCompression(final boolean enable) {
+    class MessageCompressionEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setMessageCompression(enable);
+      }
+    }
+
+    delayOrExecute(new MessageCompressionEntry());
+  }
+
+  @Override
+  public void halfClose() {
+    class HalfCloseEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.halfClose();
+      }
+    }
+
+    delayOrExecute(new HalfCloseEntry());
+  }
+
+  @Override
+  public void setAuthority(final String authority) {
+    class AuthorityEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setAuthority(authority);
+      }
+    }
+
+    delayOrExecute(new AuthorityEntry());
+  }
+
+  @Override
+  public void setDecompressorRegistry(final DecompressorRegistry decompressorRegistry) {
+    class DecompressorRegistryEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setDecompressorRegistry(decompressorRegistry);
+      }
+    }
+
+    delayOrExecute(new DecompressorRegistryEntry());
+  }
+
+  @Override
+  public void setMaxInboundMessageSize(final int maxSize) {
+    class MaxInboundMessageSizeEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setMaxInboundMessageSize(maxSize);
+      }
+    }
+
+    delayOrExecute(new MaxInboundMessageSizeEntry());
+  }
+
+  @Override
+  public void setMaxOutboundMessageSize(final int maxSize) {
+    class MaxOutboundMessageSizeEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.setMaxOutboundMessageSize(maxSize);
+      }
+    }
+
+    delayOrExecute(new MaxOutboundMessageSizeEntry());
+  }
+
+  @Override
+  public Attributes getAttributes() {
```

Comment:
This may [only be called after the application is called with onHeaders or onClose](https://github.com/grpc/grpc-java/blob/v1.7.0/core/src/main/java/io/grpc/ClientCall.java#L242). So this should simply use winningSubstream, and either throw or use EMPTY otherwise.